### PR TITLE
PLFM-4731 - Error sending messages to Synapse users re: character limitations

### PR DESCRIPTION
**Pull request by hhu94**

### CODE REVIEW

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/MessageUtils.java [hunk 1]
```diff
@@ -99,9 +99,19 @@ public static void copyDBOToDTO(DBOMessageToUser info, MessageToUser bundle) {
 		bundle.setWithUnsubscribeLink(info.getWithUnsubscribeLink());
 		bundle.setWithProfileSettingLink(info.getWithProfileSettingLink());
 		bundle.setIsNotificationMessage(info.getIsNotificationMessage());
-		bundle.setTo(info.getTo());
-		bundle.setCc(info.getCc());
-		bundle.setBcc(info.getBcc());
+		try {
+			if (info.getTo() != null) {
```
Comment (brucehoff) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/MessageUtils.java]: what if info.getTo()==null?  Should this code then call bundle.setTo(null)?

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/persistence/DBOMessageToUserBackup.java [hunk 2]
```diff
@@ -0,0 +1,231 @@
+package org.sagebionetworks.repo.model.dbo.persistence;
```
Comment (brucehoff) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/persistence/DBOMessageToUserBackup.java:1]: Please open a Jira issue to remove DBOMessageToUserBackup in the subsequent stack. (Assuming elimination of the class is part of the strategy.)

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/MessageUtils.java (hunk 1 above)
Comment (hhu94) [reply lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/MessageUtils.java]: I made the assumption that the bundle passed in had all null fields, but you're right, it's better to explicitly set the field to null.

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/persistence/DBOMessageToUser.java [hunk 3]
```diff
@@ -214,23 +226,71 @@ public MigrationType getMigratableTableType() {
...
+				} catch (UnsupportedEncodingException e) {
+					throw new RuntimeException(e);
+				}
+				return dbo;
 			}
 			
 			@Override
-			public DBOMessageToUser createBackupFromDatabaseObject(DBOMessageToUser dbo) {
-				return dbo;
+			public DBOMessageToUserBackup createBackupFromDatabaseObject(DBOMessageToUser dbo) {
```
Comment (brucehoff) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/persistence/DBOMessageToUser.java]: This doesn't help us get to the state in which the backup is DBOMessageToUser.  Is your plan that DBOMessageToUserBackup would become a permanent part of the migration scheme?

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/persistence/DBOMessageToUserBackup.java (hunk 2 above)
Comment (hhu94) [reply lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/persistence/DBOMessageToUserBackup.java:1]: Tracking that task here https://sagebionetworks.jira.com/browse/PLFM-4735

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/persistence/DBOMessageToUser.java [hunk 4]
```diff
@@ -247,8 +318,8 @@ public DBOMessageToUser createBackupFromDatabaseObject(DBOMessageToUser dbo) {
 	public int hashCode() {
 		final int prime = 31;
 		int result = 1;
-		result = prime * result + ((bcc == null) ? 0 : bcc.hashCode());
-		result = prime * result + ((cc == null) ? 0 : cc.hashCode());
+		result = prime * result + Arrays.hashCode(bytesBcc);
```
Comment (brucehoff) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/persistence/DBOMessageToUser.java]: Please let your IDE auto-generate hashcode, equals, toString (if applicable).

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/persistence/DBOMessageToUserBackup.java [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+	public byte[] getBytesBcc() {
+		return bytesBcc;
+	}
+
+	public void setBytesBcc(byte[] bytesBcc) {
+		this.bytesBcc = bytesBcc;
+	}
+
+	@Override
+	public int hashCode() {
```
Comment (brucehoff) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/persistence/DBOMessageToUserBackup.java]: regenerate hashCode and equals using IDE

File: lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/persistence/DBOMessageToUserTest.java [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+		DBOMessageToUserBackup backup = createBackup();
+		backup.setTo("Foo <foo@foo.foo>");
+		backup.setCc("Bar <bar@bar.bar>");
+		backup.setBcc("Baz <baz@baz.baz>");
+		MigratableTableTranslation<DBOMessageToUser, DBOMessageToUserBackup> translator = new DBOMessageToUser().getTranslator();
+		DBOMessageToUser translated = translator.createDatabaseObjectFromBackup(backup);
+		testNonToCcBccFieldsEqual(backup, translated);
+		assertArrayEquals(backup.getTo().getBytes("UTF-8"), translated.getBytesTo());
+		assertArrayEquals(backup.getCc().getBytes("UTF-8"), translated.getBytesCc());
+		assertArrayEquals(backup.getBcc().getBytes("UTF-8"), translated.getBytesBcc());
```
Comment (brucehoff) [inline lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/persistence/DBOMessageToUserTest.java]: add assertions that the String fields are null.

File: lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/persistence/DBOMessageToUserTest.java [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+		DBOMessageToUserBackup backup = createBackup();
+		backup.setBytesTo("Foo <foo@foo.foo>".getBytes("UTF-8"));
+		backup.setBytesCc("Bar <bar@bar.bar>".getBytes("UTF-8"));
+		backup.setBytesBcc("Baz <baz@baz.baz>".getBytes("UTF-8"));
+		MigratableTableTranslation<DBOMessageToUser, DBOMessageToUserBackup> translator = new DBOMessageToUser().getTranslator();
+		DBOMessageToUser translated = translator.createDatabaseObjectFromBackup(backup);
+		testNonToCcBccFieldsEqual(backup, translated);
+		assertArrayEquals(backup.getBytesTo(), translated.getBytesTo());
+		assertArrayEquals(backup.getBytesCc(), translated.getBytesCc());
+		assertArrayEquals(backup.getBytesBcc(), translated.getBytesBcc());
```
Comment (brucehoff) [inline lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/persistence/DBOMessageToUserTest.java]: assertions that string fields are null

Comment (brucehoff) [reply lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/persistence/DBOMessageToUserTest.java]: add test for create backup from database object.

File: lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/persistence/DBOMessageToUserTest.java [hunk 8]
```diff
@@ -16,48 +17,97 @@
...
+		// Method under test
 		DBOMessageToUser translated = translator.createDatabaseObjectFromBackup(backup);
+
 		testNonToCcBccFieldsEqual(backup, translated);
 		assertArrayEquals(backup.getTo().getBytes("UTF-8"), translated.getBytesTo());
 		assertArrayEquals(backup.getCc().getBytes("UTF-8"), translated.getBytesCc());
 		assertArrayEquals(backup.getBcc().getBytes("UTF-8"), translated.getBytesBcc());
+		assertNull(backup.getBytesTo());
+		assertNull(backup.getBytesCc());
+		assertNull(backup.getBytesBcc());
```
Comment (brucehoff) [inline lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/persistence/DBOMessageToUserTest.java]: not necessary

File: lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/persistence/DBOMessageToUserTest.java [hunk 9]
```diff
@@ -16,48 +17,97 @@
...
+		assertArrayEquals(dbo.getSubjectBytes(), translated.getSubjectBytes());
+		assertArrayEquals(dbo.getBytesTo(), translated.getBytesTo());
+		assertArrayEquals(dbo.getBytesCc(), translated.getBytesCc());
+		assertArrayEquals(dbo.getBytesBcc(), translated.getBytesBcc());
+		assertEquals(dbo.getSent(), translated.getSent());
+		assertEquals(dbo.getNotificationsEndpoint(), translated.getNotificationsEndpoint());
+		assertEquals(dbo.getProfileSettingEndpoint(), translated.getProfileSettingEndpoint());
+		assertEquals(dbo.getWithUnsubscribeLink(), translated.getWithUnsubscribeLink());
+		assertEquals(dbo.getWithProfileSettingLink(), translated.getWithProfileSettingLink());
+		assertEquals(dbo.getIsNotificationMessage(), translated.getIsNotificationMessage());
```
Comment (brucehoff) [inline lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/persistence/DBOMessageToUserTest.java:102]: check that backup string fields are null

**Comment by brucehoff** [comment]
Looks great!

**Pull request merged**